tests/app_test.py

```diff
@@ -1 +1,230 @@
+        # returned object for IDP "default":
+        {"refresh_token": fake_tokens["default"], "id_token": "eyJhbGciOiJ"},
+        # returned object for IDP "idp_a":
+        {"refresh_token": fake_tokens["idp_a"], "id_token": "eyJhbGciOiJ"},
+    ]
+    patched_fetch_access_token = mock.patch(
+        "authlib.client.OAuthClient.fetch_access_token", mocked_response
+    )
+    patched_fetch_access_token.start()
+
+    # mock `jwt.decode` to return fake data
+    now = int(time.time())
+    mocked_jwt_response = mock.MagicMock()
+    mocked_jwt_response.side_effect = [
+        # decoded id_token for IDP "default":
+        {"context": {"user": {"name": test_user.username}}},
+        # decoded refresh_token for IDP "default":
+        {"jti": str(uuid.uuid4()), "exp": now + 100, "sub": test_user.userid},
+        # decoded id_token for IDP "idp_a":
+        {"context": {"user": {"name": test_user.username}}},
+        # decoded refresh_token for IDP "idp_a":
+        {"jti": str(uuid.uuid4()), "exp": now + 100, "sub": test_user.userid},
+    ]
+    patched_jwt_decode = mock.patch("jose.jwt.decode", mocked_jwt_response)
+    patched_jwt_decode.start()
+
+    # get refresh token for IDP "default"
+    fake_state = "qwerty"
+    with client.session_transaction() as session:
+        session["state"] = fake_state
+    res = client.get(
+        "/oauth2/authorize?state={}".format(fake_state), headers=auth_header
+    )
+    assert res.status_code == 200, res.json
+
+    # get refresh token for IDP "idp_a"
+    with client.session_transaction() as session:
+        session["state"] = fake_state
+        session["idp"] = "idp_a"
+    res = client.get(
+        "/oauth2/authorize?state={}".format(fake_state), headers=auth_header
+    )
+    assert res.status_code == 200
+
+    # make sure the refresh tokens are in the DB
+    refresh_tokens = db_session.query(RefreshToken).all()
+    for t in refresh_tokens:
+        assert t.username == test_user.username
+        if t.idp == "default":
+            assert t.token == fake_tokens["default"]
+        else:
+            assert t.token == fake_tokens["idp_a"]
+
+
+def test_external_oidc_endpoint(client, test_user, db_session, auth_header):
+    with open(os.environ["SECRET_CONFIG"], "r") as f:
+        configured_oidc = json.load(f)["external_oidc"]
+    expected_oidc = {}
+    for provider in configured_oidc:
+        for idp, login_option in provider.get("login_options", {}).items():
+            expected_oidc[idp] = login_option
+            expected_oidc[idp]["base_url"] = provider["base_url"]
+            expected_oidc[idp]["oidc_client_id"] = provider["oidc_client_id"]
+
+    # GET /external_oidc before logging in
+    res = client.get("/external_oidc/", headers=auth_header)
+    assert res.status_code == 200
+    actual_oidc = res.json["providers"]
+
+    # the listed providers should be the configured providers
+    print("Configured providers: {}".format(expected_oidc))
```

Comment:
Did u mean to keep these print statements?

Reply:
yes, if the test fails we can compare the expected vs the actual data